File: backend/review_occurrences.py
"""Calendar scheduling and immutable occurrence projections (no database writes)."""
from calendar import monthrange
from datetime import datetime, timedelta

MONTHS = {"monthly": 1, "quarterly": 3, "semiannual": 6, "annual": 12}


def scheduled_date(value):
    try:
        return datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except (ValueError, TypeError):
        return None


def occurrence_id(review):
    return review.get("current_occurrence_id") or "occ_" + review["review_id"]


def schedule(review, reset_anchor=False):
    date = scheduled_date(review.get("due_date"))
    recurrence = review.get("recurrence") or "none"
    anchor = None if reset_anchor else review.get("schedule_anchor")
    anchor = anchor or ({"day": date.day, "month_end": date.day == monthrange(date.year, date.month)[1]} if date else None)
    next_date = None
    label = "Needs Scheduling"
    if date:
        label = str(date.year) if recurrence == "annual" else (
            f"Q{(date.month - 1) // 3 + 1} {date.year}" if recurrence == "quarterly" else
            f"H{(date.month - 1) // 6 + 1} {date.year}" if recurrence == "semiannual" else
            date.strftime("%B %Y") if recurrence == "monthly" else date.date().isoformat())
        if recurrence in MONTHS:
            month0 = date.month - 1 + MONTHS[recurrence]
            year, month = date.year + month0 // 12, month0 % 12 + 1
            last = monthrange(year, month)[1]
            next_date = date.replace(year=year, month=month, day=last if anchor["month_end"] else min(anchor["day"], last)).isoformat()
        elif recurrence == "custom" and isinstance(review.get("custom_recurrence_days"), int) and review["custom_recurrence_days"] > 0:
            next_date = (date + timedelta(days=review["custom_recurrence_days"])).isoformat()
    return {"period": label, "next_review_date": next_date, "schedule_anchor": anchor}


def view(review):
    result = {**review, **schedule(review), "current_occurrence_id": occurrence_id(review)}
    result.pop("_execution_lock", None)
    if review.get("status") not in ("completed", "cancelled"):
        unscheduled = not scheduled_date(review.get("due_date")) or ("recurrence" in review and review["recurrence"] in (None, "")) or (
            review.get("recurrence") not in (None, "", "none") and not result["next_review_date"])
        if unscheduled:
            result["status"] = "needs_scheduling"
        elif review.get("status") == "needs_scheduling":
            result["status"] = "upcoming"
    return result


def occurrence_query(review, selected=None, field="occurrence_id"):
    """Legacy untagged records belong ONLY to the first recorded execution."""
    selected = selected or occurrence_id(review)
    initial = "occ_" + review["review_id"]
    return {field: {"$in": [selected, None]}} if selected == initial else {field: selected}


def snapshot(review, evidence, finding_count, user, at):
    fields = ("review_id", "client_id", "title", "review_type", "due_date", "owner_id", "reviewer_id",
              "recurrence", "custom_recurrence_days", "notes", "scope", "follow_up", "policy_id", "vendor_id",
              "source", "started_by", "started_at")
    return {**{k: review.get(k) for k in fields}, **schedule(review),
            "occurrence_id": occurrence_id(review), "status": "completed", "completion_date": at,
            "completed_at": at, "completed_by": user["user_id"], "completed_by_name": user.get("name") or user.get("email"),
            "outcome": "findings_raised" if finding_count else "no_findings", "finding_count": finding_count,
            "evidence": [{k: e.get(k) for k in ("evidence_id", "filename", "version", "sha256")} for e in evidence]}

File: frontend/src/lib/reviewOccurrences.js
// Date-only obligations retain their calendar cadence, not the completion timestamp.
export const occurrenceId = review => review.current_occurrence_id || 'occ_' + review.review_id;
export function scheduledDate(value) {
  if (typeof value !== 'string' || !/^\d{4}-\d{2}-\d{2}(?:$|T| )/.test(value)) return null;
  const day = value.slice(0,10), calendar = new Date(day + 'T00:00:00Z'), date = new Date(value);
  // Date.parse normalizes February 30 instead of rejecting it.
  if (!Number.isFinite(calendar.getTime()) || calendar.toISOString().slice(0,10) !== day || !Number.isFinite(date.getTime()) || date.getUTCFullYear() < 1 || date.getUTCFullYear() > 9999) return null;
  return date;
}
export function reviewSchedule(review, resetAnchor = false) {
  const date = scheduledDate(review.due_date);
  const valid = !!date;
  const recurrence = review.recurrence || 'none';
  const anchor = (!resetAnchor && review.schedule_anchor) || (valid ? {
    day: date.getUTCDate(),
    month_end: date.getUTCDate() === new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth() + 1, 0)).getUTCDate()
  } : null);
  let next = null, period = 'Needs Scheduling';
  if (valid) {
    const year = date.getUTCFullYear(), month = date.getUTCMonth();
    period = recurrence === 'quarterly' ? `Q${Math.floor(month / 3) + 1} ${year}`
      : recurrence === 'semiannual' ? `H${Math.floor(month / 6) + 1} ${year}`
      : recurrence === 'annual' ? String(year)
      : recurrence === 'monthly' ? date.toLocaleDateString('en-US', {month:'long', year:'numeric', timeZone:'UTC'})
      : date.toISOString().slice(0,10);
    const months = {monthly:1, quarterly:3, semiannual:6, annual:12}[recurrence];
    if (months) {
      date.setUTCDate(1); date.setUTCMonth(month + months);
      const last = new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth() + 1, 0)).getUTCDate();
      date.setUTCDate(anchor.month_end ? last : Math.min(anchor.day, last));
      if (date.getUTCFullYear() <= 9999) next = date.toISOString();
    } else if (recurrence === 'custom' && Number.isInteger(review.custom_recurrence_days) && review.custom_recurrence_days > 0) {
      const candidate = new Date(date.getTime() + review.custom_recurrence_days * 86400000);
      if (!Number.isNaN(candidate.getTime()) && candidate.getUTCFullYear() <= 9999) next = candidate.toISOString();
    }
  }
  return {period, next_review_date:next, schedule_anchor:anchor};
}
export function reviewView(review) {
  const result = {...review, ...reviewSchedule(review), current_occurrence_id:occurrenceId(review)};
  if (!['completed','cancelled'].includes(review.status)) {
    const missing = !scheduledDate(review.due_date) || ('recurrence' in review && [null,''].includes(review.recurrence))
      || (!['none','',null,undefined].includes(review.recurrence) && !result.next_review_date);
    if (missing) result.status = 'needs_scheduling';
    else if (review.status === 'needs_scheduling') result.status = 'upcoming';
  }
  return result;
}
export function belongsToOccurrence(item, review, selected = occurrenceId(review)) {
  return item.occurrence_id === selected || (!item.occurrence_id && selected === 'occ_' + review.review_id);
}
export function assertCurrentOccurrence(review, selected) {
  if (!selected || selected !== occurrenceId(review) || ['completed','cancelled'].includes(review.status))
    throw new Error('This occurrence is closed or has changed. Reload the Review.');
}

// Related Findings/Actions retain their originating execution after recurrence advances.
export function relatedReviewInitialValues(review, source) {
  if (source?.review_id !== review.review_id || !source.occurrence_id) return {};
  const occurrence = review.occurrences?.find(o => o.occurrence_id === source.occurrence_id);
  return occurrence ? {occurrence} : {};
}
